query/language/reduce.feature

```diff
@@ -0,0 +1,556 @@
+
+
+  Scenario: the maximum value for a particular variable within each answer group can be retrieved using a group 'max'
+    Given connection open write transaction for database: typedb
+    Given typeql write query
+      """
+      insert
+      $c1 isa company, has name "Lloyds", has ref 0;
+      $c2 isa company, has name "Barclays", has ref 1;
+      $p1 isa person, has name "Amy", has age 48, has ref 2;
+      $p2 isa person, has name "Weiyi", has age 57, has ref 3;
+      $p3 isa person, has name "Kimberly", has age 31, has ref 4;
+      $p4 isa person, has name "Reginald", has age 45, has ref 5;
+      $e1 isa employment, links (employer: $c1, employee: $p1, employee: $p2, employee: $p3), has ref 6;
+      $e2 isa employment, links (employer: $c2, employee: $p4), has ref 7;
+      """
+    Given transaction commits
+
+    Given connection open read transaction for database: typedb
+    When get answers of typeql read query
+      """
+      match
+        $x isa company;
+        $y isa person, has age $z;
+        $r isa employment, links ($x, $y);
+      reduce $max = max($z) within ($x);
+      """
+    Then uniquely identify answer concepts
+      |     x     |    max        |
+      | key:ref:0 | value:long:57 |
+      | key:ref:1 | value:long:45 |
+
+
+  Scenario: Grouped reductions can be performed on value variables
+    Given connection open write transaction for database: typedb
+    Given typeql write query
+      """
+      insert
+        $a1 isa person, has name "Alice", has age 22, has ref 0;
+        $a2 isa person, has name "Alice", has age 18, has ref 1;
+        $b1 isa person, has name "Bob", has age 21, has ref 2;
+        $b2 isa person, has name "Bob", has age 24, has ref 3;
+      """
+    Then transaction commits
+
+    Given connection open read transaction for database: typedb
+    When get answers of typeql read query
+      """
+      match
+       $p isa person, has name $name, has age $a;
+       $n = $name;
+       $to25 = 25 - $a;
+      reduce $sum = sum($to25) within ($n);
+      """
+    Then uniquely identify answer concepts
+      |         n          |     sum       |
+      | value:string:Alice | value:long:10 |
+      | value:string:Bob   | value:long:5  |
+
+
+  Scenario: Grouped standard deviation of one value returns an empty group value
+    Given connection open schema transaction for database: typedb
+    Given typeql define
+      """
+      define
+      attribute income value double;
+      person owns income;
+      """
+    Given transaction commits
+    Given connection open write transaction for database: typedb
+    Given typeql write query
+    """
+    insert
+    $x isa person, has income 100.0, has ref 0;
+    """
+    Given transaction commits
+
+    Given connection open read transaction for database: typedb
+    When get answers of typeql read query
+      """
+      match $x isa person, has income $y;
+      select $x, $y;
+      reduce $std = std($y) within ($x);
```

Comment:
Let's require `$std? = std(y) within $x`
Basically, whenever the reducer is falliable, we must use an optionally-valued variable on the left.

Reply:
It doesn't parse, but I've updated it.